File: my-folder/0025-reverse-nodes-in-k-group/solution.cpp
class Solution {
public:
    int len(ListNode* head){
        int length=0;
        while(head){
            head=head->next;
            length++;
        }
        return length;
    }
    
    ListNode* reverse(ListNode* head){
        ListNode* current = head;
        ListNode* prev=nullptr;
        while(current){
            ListNode* tail = current->next;
            current->next=prev;
            prev=current;
            current=tail;
        }
        return prev; 
    }
    
    ListNode* reverseKGroup(ListNode* head, int k) {
        if(!head) return head;
        int length = len(head);
        int groups = length / k;
        ListNode* dummy = new ListNode(-1);
        dummy->next=head;
        ListNode* newH=dummy;
        while(groups--){
            ListNode* current = newH->next;
            ListNode* tail = current;
            for(int i=0;i<k-1 && tail;i++){
                tail=tail->next;
            }
            if(!tail) break;
            ListNode* nextH=tail->next;
            tail->next=nullptr;
            ListNode* reverseH=reverse(current);
            newH->next=reverseH;
            current->next=nextH;
            newH=current;


        }
        return dummy->next;
        
        
    }
};

File: my-folder/0025-reverse-nodes-in-k-group/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    ListNode reverse(ListNode head){
        ListNode current = head;
        ListNode prev = null;
        while(current!=null){
            ListNode temp = current.next;
            current.next = prev;
            prev = current;
            current = temp;
        }
        return prev;
    }
    int lenList(ListNode head){
        int length = 0;
        while(head!=null){
            head=head.next;
            length++;
        }
        return length;

    }
    public ListNode reverseKGroup(ListNode head, int k) {
        if(head==null || k==1) return head;
        int groups = lenList(head)/k;
        ListNode dummy = new ListNode(-1);
        ListNode newH = dummy;
        dummy.next = head;
        while(groups!=0){
            ListNode startH = newH.next;
            ListNode current = startH;
            for(int i=1;i<k;i++){
                current = current.next;
            }
            if(current==null) break;
            ListNode nextH=current.next;
            current.next = null;
            newH.next = reverse(startH);
            startH.next = nextH;
            newH=startH;
            groups--;
        }
        return dummy.next;
       
    }
}
